### PR TITLE
[BC-373] Align Gossip message ID according to the spec

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- Align Gossip `messageId` according to the spec: https://github.com/ethereum/eth2.0-specs/blob/v0.11.1/specs/phase0/p2p-interface.md#topics-and-messages

Depends on #1680
Pending for Libp2p `0.3.4` release to be published

### CODE REVIEW

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.jetbrains.annotations.NotNull;
+import pubsub.pb.Rpc.Message;
+
+/** Customization of a standard Libp2p Gossip router for Eth2 spec */
+public class Eth2GossipRouter extends GossipRouter {
+
+  @NotNull
+  @Override
+  protected String getMessageId(@NotNull Message msg) {
+    return Base64.encodeBytes(Hash.sha2_256(msg.getData().toByteArray()));
```
Comment (ajsutton) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java]: I think this is the wrong base64.  It should be the URL safe variant without padding.  To do that in discv5 I had to use: https://github.com/PegaSysEng/discovery/blob/247107e61cd53ed4c05ea2401985094ae1c3668d/src/main/java/org/ethereum/beacon/discovery/schema/NodeRecord.java#L84

**Comment by ajsutton** [comment]
One other thought - should we be rejecting gossip messages that don't set the message ID correctly?

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java (hunk 1 above)
Comment (Nashatyrev) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java]: Oh, thanks! Absolutely my bad, should read the spec more carefully. 
Didn't even think there might be 'wrong' base64 :)

Comment (Nashatyrev) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java]: BTW, wouldn't this work? 
```java
java.util.Base64.getUrlEncoder().withoutPadding()
```
UPD: just checked - these are equivalent

**Comment by Nashatyrev** [comment]
> One other thought - should we be rejecting gossip messages that don't set the message ID correctly?

Yep interesting question. The `messageId`s are only exchanged with `IHAVE, IWANT` gossip messages. You can't even detect that remote peer uses wrong `messageId`. 

For example your (`peer1`) and remote peer (`peer2`) both have the message `M`, but your peer calculates its id as `ID1` and remote as `ID2`. So remote peer would be sending you 0..N times the message `IHAVE(ID2)` your peer doesn't find it and request `IWANT(ID2)`, then the remote peer would send you message `M` (which doesn't contain `messageId` in its structure), you would just see it's a duplicate and skip it. And vice versa you would be sending remote peer `IHAVE(ID1)` and so on... 

So this is generally undetectable situation which doesn't affect network liveness but just may increase its load

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java (hunk 1 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/Eth2GossipRouter.java]: I did not know that method existed.... https://github.com/PegaSysEng/discovery/pull/43 will switch discovery to use it.

**Review by ajsutton — APPROVED** [review]
LGTM.

**Pull request merged**